src/hid/hid_device.cc

```diff
@@ -354,6 +355,60 @@ Status HidDevice::ReceiveCommand(absl::Duration timeout, uint8_t* cmd,
   return Status::kErrNone;
 }
 
+Status HidDevice::SendRecvCommand(uint8_t send_cmd,
+                                  const std::vector<uint8_t>& send_data,
+                                  std::vector<uint8_t>* recv_data,
+                                  bool expect_response, bool expect_status,
+                                  bool expect_empty_data,
+                                  bool expect_up_check) const {
+  OK_OR_RETURN(SendCommand(send_cmd, send_data));
+  uint8_t recv_cmd;
+  Status recv_status = ReceiveCommand(kReceiveTimeout, &recv_cmd, recv_data);
+  if (!expect_response && recv_status != Status::kErrTimeout) {
+    tracker_->AddObservation(
+        "The command doesn't expect response but got one.");
+  }
+  if (recv_status != Status::kErrNone) {
+    return recv_status;
+  }
+  // The answer might also be a keepalive.
+  bool has_sent_prompt = false;
+  while (recv_cmd == kCtapHidKeepalive) {
+    KeepaliveStatus keepalive_response = ProcessKeepalive(*recv_data);
+    if (keepalive_response == KeepaliveStatus::kStatusError)
+      return Status::kErrOther;
+    if (keepalive_response == KeepaliveStatus::kStatusUpNeeded &&
+        !has_sent_prompt) {
+      has_sent_prompt = true;
+      if (!tracker_->IsTouchPromptIgnored()) {
+        PromptUser();
+      }
+    }
+    OK_OR_RETURN(ReceiveCommand(kReceiveTimeout, &recv_cmd, recv_data));
+  }
+
+  if (has_sent_prompt && !expect_up_check) {
+    tracker_->AddObservation("A prompt was sent unexpectedly.");
+  }
+  if (!has_sent_prompt && expect_up_check) {
+    tracker_->AddObservation(
+        "A prompt was expected, but not performed. Sometimes it is just not "
```

Comment:
This function's structure seems to be:
`SendCommand`
`ReceiveCommand`
expect response?
_Lots of code after receiving that you probably want to reuse._
And then you handle all cases with your bools.

Here's an idea: Instead of a helper function that handles all cases, can we go back to all functions doing that themselves, and writing a different helper function for the reused part instead? Then you can call this `KeepaliveHelper` (please choose a better name) wherever you need it. It would mean that you can't just call `SendRecvCommand` with lots of `bool`s everywhere, but the hidden complexity in understand all implications of these parameters might still make it more readable to have this little bit of code duplication.